libs/database/src/lib/services/authentication.service.v2.ts

```diff
@@ -143,4 +157,188 @@ export class AuthenticationServiceV2 {
       pages: pages.map((page) => page.toPageResponse()),
     };
   }
+
+  /**
+   *
+   * @param token
+   * @returns RefreshTokenResponse
+   */
+  async getRefreshToken(token: string) {
+    const credential = await this.getCredentialFromRefreshToken(token);
+    if (!(credential && credential.isRefreshTokenValid()))
+      throw CastcleException.INVALID_REFRESH_TOKEN;
+
+    const userProfile: {
+      profile: User;
+      pages: {
+        items: User[];
+        pagination: Pagination;
+      };
+    } = await this.getUserProfile(credential);
+
+    this.logger.log('Validate profile member.');
+
+    if (!userProfile.profile && !credential.account.isGuest) {
+      this.logger.warn('Member Profile is empty.');
+      throw CastcleException.INVALID_REFRESH_TOKEN;
+    }
+
+    const refreshToken = await this.mapRefreshToken(credential, userProfile);
+    return refreshToken;
+  }
+  /**
+   *
+   * @param refreshToken refreshToken for find in credential
+   * @returns result of finding with refreshToken
+   */
+  private getCredentialFromRefreshToken = (refreshToken: string) =>
+    this.credentialModel.findOne({ refreshToken: refreshToken }).exec();
+
+  /**
+   *
+   * @param credential credential
+   * @param userProfile userProfileWithPageCredential
+   * @returns RefreshTokenResponse
+   */
+  private async mapRefreshToken(
+    credential: Credential,
+    userProfile: {
+      profile: User;
+      pages: {
+        items: User[];
+        pagination: Pagination;
+      };
+    }
+  ) {
+    const accessTokenPayload = await this.getAccessTokenPayloadFromCredential(
+      credential
+    );
+    const renewAccessToken = this.getRenewAccessToken(
+      credential,
+      accessTokenPayload
+    );
+    const getAccount = this.getAccountFromId(credential.account._id);
+
+    const [accessToken, account] = await Promise.all([
+      renewAccessToken,
+      getAccount,
+    ]);
+
+    const profile = userProfile.profile
+      ? await userProfile.profile.toUserResponse({
+          passwordNotSet: account.password ? false : true,
+          mobile: account.mobile,
+        })
+      : null;
+
+    const pages = userProfile.pages
+      ? userProfile.pages.items.map((item) => item.toPageResponse())
+      : null;
+
+    const response: RefreshTokenResponse = {
+      profile,
+      pages,
+      accessToken,
+    };
+    return response;
+  }
+
+  /**
+   *
+   * @param payload AccessTokenPayload | MemberAccessTokenPayload
+   * @returns
+   */
+  private async getRenewAccessToken(
+    credential: Credential,
+    payload: AccessTokenPayload | MemberAccessTokenPayload
+  ) {
+    const result = this.generateAccessToken(payload);
```

Comment:
ระบุชื่อ variable ให้สื่อความหมาย ไม่ควรใช้ result or data
```suggestion
    const newAccessToken = this.generateAccessToken(payload);
```